fms_extras/utils/generation.py

```diff
@@ -3,12 +3,344 @@
 from typing import Any, Callable, Dict, List, MutableMapping, Optional, Tuple, Union
 
 import torch
+import torch.nn as nn
 import torch.nn.functional as F
 
 from fms_extras.models.speculator import MLPSpeculator, apply_index_map, flatten_batch
 from fms_extras.utils.cache.paged import PagedAttentionCacheData, PagedKVCacheManager
 
 
+def __prepare_inputs_for_prefill(
+    input_ids: Union[torch.Tensor, List[torch.Tensor]],
+    kv_cache_manager: PagedKVCacheManager,
+    model_input_lengths: List[int],
+    max_length_in_batch: int,
+    decode_seq_length: int,
+    max_sequence_length_in_model: int,
+) -> Tuple[torch.Tensor, PagedAttentionCacheData, torch.Tensor]:
+    """Prepare all inputs required to perform prefill step on prompt tokens
+
+    This function will:
+
+    Args:
+        input_ids: Union[torch.Tensor, List[torch.Tensor]]
+            list of prompts where each tensor in the list corresponds to a single prompt in the batch
+        kv_cache_manager: PagedKVCacheManager
+            the paged attention kv-cache manager
+        model_input_lengths: List[int]
+            list where each value corresponds the number of tokens in the prompt at that index
+        max_length_in_batch: int
+            the max prompt length in the batch
+        decode_seq_length: int
+            the model input length at decode step
+        max_sequence_length_in_model: int
+            the max sequence length given to generate, typically the model's max sequence length
+
+    Returns:
+    Tuple[torch.Tensor, PagedAttentionCacheData, torch.Tensor]
+        the padded input ids, the cache data, and the mask
+    """
+    # pad the inputs
+    inputs = torch.stack(
+        [
+            F.pad(input_ids[i], (max_length_in_batch - model_input_lengths[i], 0))
+            for i in range(len(input_ids))
+        ]
+    )
+
+    # reserve blocks in the cache for the prompts
+    cache_data: PagedAttentionCacheData = kv_cache_manager.allocate_tokens(
+        model_input_lengths
+    )
+
+    # Build padded causal mask
+    mask = __create_prefill_mask(model_input_lengths, inputs.device)
+
+    # get the last tokens if past max length in model
+    inputs = inputs[:, -max_sequence_length_in_model + decode_seq_length :]
```

Comment:
Edge case issue - if max length of inputs is longer that the model max sequence length, then we need to truncate the inputs AND the cache data and mask. We should perform this step earlier and avoid creating overlarge objects in the first place (does this also affect the `model_input_length` and `max_length_in_batch` in `speculative_generate`?). This was probably an uncaught edge case in the older code too...